![Les billets de Semaine52 / AuFilDuBoamp / DataMining des données de marchés publics au Bulletin officiel](https://raw.githubusercontent.com/Semaine52/AuFilDuBoamp/master/IMG/BILLETS_S52_AUFD_BANDEAU.jpg "Logo Title Text 1")

##### Data Mining des données de la commande publique parues au **Bulletin officiel des annonces des marchés publics (BOAMP 2015-2019)**
Source open data : https://www.data.gouv.fr/fr/datasets/api-boamp-beta/
# **Aujourd'hui : la Ville Fantôme de Cedex et autres Joyeusetés**
### *ou «Les ravages de l'absence de référentiel. [...]» (@JoelGombin, Twitter, 19/09/2019)*

____

# Constatations

#### Qui ne s'est jamais demandé ce qu'il s'est passé à Toulouse au BOAMP en 2018?   
#### Pour commencer, vous récupérez toute la colonne «ville» extraite des annonces de l'année.
#### Cela donne ceci:

In [1]:
# IMPORTS DES OUTILS
from urllib.request import urlretrieve
import pandas as pd

# RÉGLAGE IPYTHON INTERACTVESHELL
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all" 

# RÉGLAGES PANDAS
pd.set_option('display.notebook_repr_html', True)
pd.set_option('display.max_colwidth', -1)
pd.set_option('display.max_rows', 2000)
pd.set_option('display.max_columns', 500)

# IMPORT ET LECTURE DES DONNÉES
URL = 'https://www.semaine52.com/AuFilDuBoamp/PKL/df_d01_08_ville_2018_no_index.pkl'
urlretrieve(URL, 'df_d01_08_ville_2018_no_index.pkl');
df = pd.read_pickle('df_d01_08_ville_2018_no_index.pkl')
a, _ = df.shape

print('*' * 100)
print(f"Nombre d'annonce au BOAMP en 2018: {a} ")
print(f"Ci-dessous, les cinq premières lignes du tableau de résultats:")

df.head()

****************************************************************************************************
Nombre d'annonce au BOAMP en 2018: 155935 
Ci-dessous, les cinq premières lignes du tableau de résultats:


d01_08_ville
0  Magny-les-Hameaux    
1  Conflans Ste-Honorine
2  ANNECY               
3  Toulouse cedex 1     
4  Caudry

____

#### Toujours en confiance, à partir des 115 935 entrées de départ, vous réduisez votre recherche à Toulouse:

In [2]:
df_toulouse = df.loc[lambda df: df.d01_08_ville == 'Toulouse', :].reset_index(drop=True)
df_toulouse.shape
df_toulouse.head()

(795, 1)

d01_08_ville
0  Toulouse   
1  Toulouse   
2  Toulouse   
3  Toulouse   
4  Toulouse

#### Vous auriez donc 795 annonces sur l'année 2018 à Toulouse en stock.
#### Mais il y a un problème, on le voit plus haut, c'est le fameux «cedex».

#### Je vous épargne ici les laborieuses étapes de nettoyage (qui feront l'objet d'un autre article), pour vous montrer ceci:



In [3]:
df_boamp_2018_classe_par_nom_de_ville = df.sort_values(by='d01_08_ville').reset_index(drop=True)
df_boamp_2018_classe_par_nom_de_ville.iloc[0:10]

d01_08_ville
0  01330 Villars les Dombes      
1  06364, NICE CEDEX 4           
2  10120                         
3  10400 Nogent sur seine        
4  113Lyon Saint Exupéry Aéroport
5  11890 Carcassonne cedex       
6  11890 Carcassonne cedex       
7  12101 Millau                  
8  18000 Bourges                 
9  21033 dijon cedex

#### Ce qu'on y voit, c'est l'implémentation par l'acheteur de la colonne «ville» quand on classe la liste par ordre alphabétique.
#### Si vous lisez cet article sur JupyterLab (voir plus loin), je vous invite à vous balader un peu plus loin dans la base, c'est parfois cocasse...


___

#### Je vous épargne la recherche de la première ville correctement entrée: c'est «ABLIS» en position ... 117.

In [5]:
df_boamp_2018_classe_par_nom_de_ville.iloc[117,:]

d01_08_ville    ABLIS
Name: 117, dtype: object

___

# Conséquences

Tous les jeux de données sont à remanier, nettoyer, etc.  C'est comme ça.

Toujours est-il que ce nettoyage est plus ou moins compliqué à faire:
- Il faut unifier les casses;
- Faire attention aux caractères spéciaux;
- Se débarrasser des espaces superflux;
- etc.

Et après seulement, on peut espérer travailler sur les données.

À titre d'exemple, si vous souhaitez géocoder vos annonces à partir de cette colonne, toutes les entrées avec un cedex ou autre décoration sera ignorée. 
Vous pourrez vous rabattre sur les codes postaux, mais vous y rencontrerez sans doute des problèmes similaires.

# Conclusion

L'usage d'un bon référentiel d'implémentation de la base fait défault au BOAMP.

Or, des données de marchés publics rendues (encore) plus absconses sont rapidement inutilisables.  

La cible de transparence affichée dans les textes par le législateur, voire le constituant, est de ce fait manquée.

# Aller plus loin avec ces données

### Pour m'éviter d'être assommant avec un trop plein de lignes de code dans cet article, je n'ai pas poussé les manipulations avec le jeu de données.
### Si vous voulez vous amuser de votre côté sans avoir à télécharger tout un environnement, voici quoi faire:
- 1// Rendez-vous sur Binder à cette adresse : https://mybinder.org/v2/gh/Semaine52/Semaine52Lab/master?urlpath=lab/tree/CONSOLE_PYTHON_A.ipynb
- 2// Copiez/Collez/Exécutez le snippet ci-dessous dans une cellule:



